src/visitors/sympy_replace_solutions_visitor.cpp

```diff
@@ -0,0 +1,354 @@
+            !solutions_.is_all_untagged()) {
+            logger->warn(
+                "SympyReplaceSolutionsVisitor :: not all solutions were replaced. Adding remaining "
+                "statements at the end. Probably something went wrong.");
+        }
+
+        pre_solve_statements_.emplace_back_all_statements(new_statements, true);
+        tmp_statements_.emplace_back_all_statements(new_statements, true);
+        solutions_.emplace_back_all_statements(new_statements, true);
+    }
+
+    node.set_statements(std::move(new_statements));
+}
+
+void SympyReplaceSolutionsVisitor::visit_diff_eq_expression(ast::DiffEqExpression& node) {
+    interleaves_counter_.new_equation(true);
+
+    const auto& statement = std::static_pointer_cast<ast::Statement>(
+        node.get_parent()->get_shared_ptr());
+
+    // do not visit if already marked
+    if (replacements_.find(statement) != replacements_.end()) {
+        return;
+    }
+
+    switch (policy_) {
+    case ReplacePolicy::VALUE: {
+        const auto dependencies = statement_dependencies(node.get_expression()->get_lhs(),
+                                                         node.get_expression()->get_rhs());
+        const auto& key = dependencies.first;
+        const auto& vars = dependencies.second;
+
+        if (solutions_.is_var_assigned_here(key)) {
+            ast::StatementVector new_statements;
+
+            pre_solve_statements_.emplace_back_all_statements(new_statements);
+            tmp_statements_.emplace_back_all_statements(new_statements);
+            solutions_.try_emplace_back_statement(new_statements, key);
+
+            replacements_.emplace(statement, new_statements);
+        }
+        break;
+    }
+    case ReplacePolicy::GREEDY: {
+        if (!solutions_.is_all_untagged()) {
+            ast::StatementVector new_statements;
+
+            pre_solve_statements_.emplace_back_all_statements(new_statements);
+            tmp_statements_.emplace_back_all_statements(new_statements);
+            solutions_.emplace_back_next_statement(new_statements);
+
+            replacements_.emplace(statement, new_statements);
+        }
+        break;
+    }
+    }
+}
+
+void SympyReplaceSolutionsVisitor::visit_lin_equation(ast::LinEquation& node) {
+    interleaves_counter_.new_equation(true);
+
+    const auto& statement = std::static_pointer_cast<ast::Statement>(
+        node.get_parent()->get_shared_ptr());
+
+    const auto dependencies = statement_dependencies(node.get_left_linxpression(),
+                                                     node.get_linxpression());
+
+    // do not visit if already marked
+    if (replacements_.find(statement) != replacements_.end()) {
+        return;
+    }
+
+    switch (policy_) {
+    case ReplacePolicy::VALUE: {
+        const auto dependencies = statement_dependencies(node.get_left_linxpression(),
+                                                         node.get_linxpression());
```

Comment:
Why do you do this after you already get `statement_dependenices` on line 165-166?